### PR TITLE
Internal ppm 2306

### PR DESCRIPTION
**Pull request by cmarzer**
Tested on my modified taranis with internal jeti module. The internal ppm seemed  to work perfectly well. I just added the GUI functions. One should check if things still work well with the original frsky internal module (I don't have any unmodified transmitter to test that part).

### CODE REVIEW

File: radio/src/Makefile [hunk 1]
```diff
@@ -336,10 +336,9 @@ TRACE_AUDIO = NO
 LCD_DUAL_BUFFER = NO
 
 # Enable internal module PPM mode for Taranis
-# Notice: enabling this only enable code in the driver,
-# the menu selection is still not possible.
+
 # Values = NO, YES
-TARANIS_INTERNAL_PPM = NO
+TARANIS_INTERNAL_PPM = YES
```
Comment (bsongis) [inline radio/src/Makefile]: No!

File: radio/src/eeprom_conversions.cpp [hunk 2]
```diff
@@ -567,6 +567,7 @@ PACK(typedef struct {
 #if defined(PCBTARANIS)
 #define MODELDATA_EXTRA_216 \
   uint8_t externalModule; \
+  uint8_t internalModule; \
```
Comment (bsongis) [inline radio/src/eeprom_conversions.cpp]: You introduce a shift in old EEPROM?

File: radio/src/gui/Taranis/menu_model_setup.cpp [hunk 3]
```diff
@@ -216,8 +220,22 @@ void menuModelSetup(uint8_t event)
     #define POT_WARN_ITEMS()                uint8_t(g_model.potsWarnMode ? NAVIGATION_LINE_BY_LINE|NUM_POTS : 0)
   #endif
   bool CURSOR_ON_CELL = (m_posHorz >= 0);
+#if defined(TARANIS_INTERNAL_PPM)
+  MENU_TAB({ 0, 0, TIMERS_ROWS, 0, 1, 0, 0, LABEL(Throttle), 0, 0, 0, LABEL(PreflightCheck), 0, 0, SW_WARN_ITEMS(), POT_WARN_ITEMS(), NAVIGATION_LINE_BY_LINE|(NUM_STICKS+NUM_POTS+NUM_ROTARY_ENCODERS-1), 0, 
+LABEL(InternalModule),
```
Comment (bsongis) [inline radio/src/gui/Taranis/menu_model_setup.cpp]: Yes, good idea to split this into many lines, but an indentation is needed

File: radio/src/gui/Taranis/menu_model_setup.cpp [hunk 4]
```diff
@@ -570,9 +588,40 @@ void menuModelSetup(uint8_t event)
         if (attr) g_model.noGlobalFunctions = !checkIncDecModel(event, !g_model.noGlobalFunctions, 0, 1);
         break;
 
-      case ITEM_MODEL_INTERNAL_MODULE_LABEL:
```
Comment (bsongis) [inline radio/src/gui/Taranis/menu_model_setup.cpp]: Indentation is wrong in this file. We use 2 chars here.

File: radio/src/gui/Taranis/menu_model_setup.cpp [hunk 5]
```diff
@@ -570,9 +588,40 @@ void menuModelSetup(uint8_t event)
...
-        lcd_putsLeft(y, TR_INTERNALRF);
+       case ITEM_MODEL_INTERNAL_MODULE_LABEL:
```
Comment (bsongis) [inline radio/src/gui/Taranis/menu_model_setup.cpp]: The code below is impossible to read, perhaps you use tabs with tab != 8 chars which renders very badly (non standard). We use 2 chars in this project.

File: radio/src/myeeprom.h [hunk 6]
```diff
@@ -448,7 +448,7 @@ PACK(typedef struct {
   };
   #define IS_TRAINER_EXTERNAL_MODULE() (g_model.trainerMode == TRAINER_MODE_MASTER_SBUS_EXTERNAL_MODULE || g_model.trainerMode == TRAINER_MODE_MASTER_CPPM_EXTERNAL_MODULE)
   #define MODELDATA_BITMAP  char bitmap[LEN_BITMAP_NAME];
-  #define MODELDATA_EXTRA   uint8_t externalModule:3; uint8_t trainerMode:3; uint8_t potsWarnMode:2; ModuleData moduleData[NUM_MODULES+1]; char curveNames[MAX_CURVES][6]; ScriptData scriptsData[MAX_SCRIPTS]; char inputNames[MAX_INPUTS][LEN_INPUT_NAME]; uint8_t potsWarnEnabled; int8_t potsWarnPosition[NUM_POTS];
+  #define MODELDATA_EXTRA   uint8_t externalModule:3; uint8_t internalModule:3; uint8_t trainerMode:3; uint8_t potsWarnMode:2; ModuleData moduleData[NUM_MODULES+1]; char curveNames[MAX_CURVES][6]; ScriptData scriptsData[MAX_SCRIPTS]; char inputNames[MAX_INPUTS][LEN_INPUT_NAME]; uint8_t potsWarnEnabled; int8_t potsWarnPosition[NUM_POTS];
```
Comment (bsongis) [inline radio/src/myeeprom.h]: Size for this bitfield becomes strange!

File: radio/src/opentx.h [hunk 7]
```diff
@@ -462,14 +462,19 @@ extern void boardInit();
 #endif
 
 #if defined(PCBTARANIS)
+ #if defined(TARANIS_INTERNAL_PPM)
```
Comment (bsongis) [inline radio/src/opentx.h]: Indentation here as well

File: radio/src/opentx.h [hunk 8]
```diff
@@ -462,14 +462,19 @@ extern void boardInit();
...
+  #define IS_MODULE_PPM(idx)                (idx==TRAINER_MODULE || (idx==EXTERNAL_MODULE && g_model.externalModule==MODULE_TYPE_PPM)|| (idx==INTERNAL_MODULE && g_model.internalModule==MODULE_TYPE_PPM))
```
Comment (bsongis) [inline radio/src/opentx.h]: Cosmetics, but I would write INTERNAL MODULE part before EXTERNAL MODULE

File: radio/src/opentx.h [hunk 9]
```diff
@@ -462,14 +462,19 @@ extern void boardInit();
...
+  #define IS_MODULE_XJT(idx)                (((idx==INTERNAL_MODULE && g_model.internalModule==MODULE_TYPE_XJT)|| (idx==EXTERNAL_MODULE && g_model.externalModule==MODULE_TYPE_XJT)) && (g_model.moduleData[idx].rfProtocol != RF_PROTO_OFF))
+  #else
   #define IS_MODULE_PPM(idx)                (idx==TRAINER_MODULE || (idx==EXTERNAL_MODULE && g_model.externalModule==MODULE_TYPE_PPM))
-  #define IS_MODULE_XJT(idx)                ((idx==INTERNAL_MODULE || g_model.externalModule==MODULE_TYPE_XJT) && (g_model.moduleData[idx].rfProtocol != RF_PROTO_OFF))
+  #define IS_MODULE_XJT(idx)                (((idx==INTERNAL_MODULE && g_model.internalModule==MODULE_TYPE_XJT)|| (idx==EXTERNAL_MODULE && g_model.externalModule==MODULE_TYPE_XJT)) && (g_model.moduleData[idx].rfProtocol != RF_PROTO_OFF))
```
Comment (bsongis) [inline radio/src/opentx.h]: Why did you change this test?

Comment (cmarzer) [inline radio/src/opentx.h]: It is used in menu_model_setup.cpp (MENU_TAB) and internal module doesn't necessarly mean XJT anymore as before. Or did I miss something ?

**Comment by bsongis** [comment]
Next time would you write something like Fixes #issue number, it will close the issue automatically and add double links between the commit and the issue! Thanks

File: radio/src/myeeprom.h (hunk 6 above)
Comment (cmarzer) [inline radio/src/myeeprom.h]: Yeah. One could add extra bits. How do you prefer to have it fixed ?

File: radio/src/gui/Taranis/menu_model_setup.cpp (hunk 5 above)
Comment (cmarzer) [inline radio/src/gui/Taranis/menu_model_setup.cpp]: That is good to know. Is there a document about coding style available ?

File: radio/src/opentx.h (hunk 9 above)
Comment (bsongis) [inline radio/src/opentx.h]: You added an #ifdef for this and if I am not mistaken we are in the #else here!

File: radio/src/myeeprom.h (hunk 6 above)
Comment (bsongis) [inline radio/src/myeeprom.h]: If the count is correct, I see 11bits, that's uncommon

File: radio/src/opentx.h (hunk 9 above)
Comment (bsongis) [inline radio/src/opentx.h]: Yes you changed the #else part, so you possibly break something for people who don't have your mod.

**Comment by cmarzer** [comment]
The code should be clean now. I couldn't find any issue because of the TARANIS_INTERNAL_PPM code. If I understand it well, it doesn't use any DMA stream, just interrupts and its timer that isn't used anywhere else if I am not mistaken...but I will continue to test it carefully...

**Comment by bsongis** [comment]
Sorry for the remarks, you made further changes. At the end you will squash all commits into one, ok?

File: radio/src/gui/Taranis/menu_model_setup.cpp [hunk 10]
```diff
@@ -216,8 +220,23 @@ void menuModelSetup(uint8_t event)
     #define POT_WARN_ITEMS()                uint8_t(g_model.potsWarnMode ? NAVIGATION_LINE_BY_LINE|NUM_POTS : 0)
   #endif
   bool CURSOR_ON_CELL = (m_posHorz >= 0);
+#if defined(TARANIS_INTERNAL_PPM)
+  MENU_TAB({ 0, 0, TIMERS_ROWS, 0, 1, 0, 0, LABEL(Throttle), 0, 0, 0, LABEL(PreflightCheck), 0, 0, SW_WARN_ITEMS(), POT_WARN_ITEMS(),
```
Comment (bsongis) [inline radio/src/gui/Taranis/menu_model_setup.cpp]: Perfect now

File: radio/src/myeeprom.h [hunk 11]
```diff
@@ -467,7 +467,11 @@ PACK(typedef struct {
   };
   #define IS_TRAINER_EXTERNAL_MODULE() (g_model.trainerMode == TRAINER_MODE_MASTER_SBUS_EXTERNAL_MODULE || g_model.trainerMode == TRAINER_MODE_MASTER_CPPM_EXTERNAL_MODULE)
   #define MODELDATA_BITMAP  char bitmap[LEN_BITMAP_NAME];
-  #define MODELDATA_EXTRA   uint8_t externalModule:3; uint8_t trainerMode:3; uint8_t potsWarnMode:2; ModuleData moduleData[NUM_MODULES+1]; char curveNames[MAX_CURVES][6]; ScriptData scriptsData[MAX_SCRIPTS]; char inputNames[MAX_INPUTS][LEN_INPUT_NAME]; uint8_t potsWarnEnabled; int8_t potsWarnPosition[NUM_POTS];
+  #if defined(TARANIS_INTERNAL_PPM)
+    #define MODELDATA_EXTRA   uint8_t externalModule:3; uint8_t internalModule:3; uint8_t trainerMode:3; uint8_t potsWarnMode:2; ModuleData moduleData[NUM_MODULES+1]; char curveNames[MAX_CURVES][6]; ScriptData scriptsData[MAX_SCRIPTS]; char inputNames[MAX_INPUTS][LEN_INPUT_NAME]; uint8_t potsWarnEnabled; int8_t potsWarnPosition[NUM_POTS];
```
Comment (bsongis) [inline radio/src/myeeprom.h]: Still not good!!!
The EEPROM changes according to the user options!
And again a bitfield with 11bits is not good, please find some free bits somewhere else!

Comment (bsongis) [inline radio/src/myeeprom.h]: Why not in ModuleData?

**Comment by bsongis** [comment]
I changed the ModuleData structure so that you can use it to store internal module settings (on 4bits)

**Comment by projectkk2glider** [comment]
Just adding reference to original issue #2306

File: radio/src/Makefile [hunk 12]
```diff
@@ -340,8 +340,7 @@ TRACE_AUDIO = NO
 LCD_DUAL_BUFFER = NO
 
 # Enable internal module PPM mode for Taranis
-# Notice: enabling this only enable code in the driver,
-# the menu selection is still not possible.
+
```
Comment (bsongis) [inline radio/src/Makefile]: No white line here

File: radio/src/opentx.h [hunk 13]
```diff
@@ -462,14 +462,23 @@ extern void boardInit();
 #endif
 
 #if defined(PCBTARANIS)
-  #define IS_MODULE_PPM(idx)                (idx==TRAINER_MODULE || (idx==EXTERNAL_MODULE && g_model.moduleData[EXTERNAL_MODULE].type==MODULE_TYPE_PPM))
-  #define IS_MODULE_XJT(idx)                ((idx==INTERNAL_MODULE || g_model.moduleData[EXTERNAL_MODULE].type==MODULE_TYPE_XJT) && (g_model.moduleData[idx].rfProtocol != RF_PROTO_OFF))
+  #if defined(TARANIS_INTERNAL_PPM)
+    #define IS_MODULE_PPM(idx)                (idx==TRAINER_MODULE || (idx==INTERNAL_MODULE && g_model.moduleData[INTERNAL_MODULE].type==MODULE_TYPE_PPM)|| (idx==EXTERNAL_MODULE && g_model.moduleData[EXTERNAL_MODULE].type==MODULE_TYPE_PPM))
```
Comment (bsongis) [inline radio/src/opentx.h]: Alignment not right

File: radio/src/pulses/pulses_arm.cpp [hunk 14]
```diff
@@ -53,7 +53,21 @@ void setupPulses(unsigned int port)
   switch (port) {
 #if defined(PCBTARANIS)
     case INTERNAL_MODULE:
+#if defined(TARANIS_INTERNAL_PPM)
+	switch (g_model.moduleData[INTERNAL_MODULE].type) {
```
Comment (bsongis) [inline radio/src/pulses/pulses_arm.cpp]: Indentation problem

**Comment by bsongis** [comment]
All good except the little cosmetics problems!

**Comment by bsongis** [comment]
Perfect!
